Compatibility with ServerApp backend

pdstack opened the pull request:
_edit_ I have realised this is actually a ServerApp backend compatibility issue. See later post in this thread for update.

Appmode doesn't work with newer versions of jupyter_client, this pull request attempts to resolve these issuses. A temp_dir option is also added to allow for custom app mode directories, within the current jupyter path.

There are a number of changes in this pull request that may require discussion or further work, happy to amend if there are better ways of doing this.

Compatibility changes:
- Newer versions of the content manager do not allow for hidden files unless an option is set, which then shows all hidden files. To resolve this the temporary notebooks are no longer hidden.
- Content manager and session manager are now async, so check for Awaitables and await them if present. Removed older style tornado.gen and replaced with native async functions. This means it is no longer Python 2 compatible. Could rewrite this to maintain Python 2 compatibility. This still works with older versions of jupyter_client.
- Launching notebooks using command line `jupyter notebook` is now causing tornado exceptions when interacting with widgets. This may be fixed in future, but it is preferable to work with the `jupyter lab` command anyway, which doesn't have this problem. The `jupyter lab` command results in the notebook.html and page.html templates not being served and available for appmode to use, to get around this I have copied the latest version of these templates from nbclassic into the repo, with very small changes. This is an ugly way of fixing this problem, if anyone has a better idea on how to do this please let me know.

oschuett commented:
Thank you so much for this pull request!

The `server_extension.py` looks good.

I'm just a bit worried about forking the templates. Perhaps there is a better way. What do you think @unkcpz?

pdstack commented:
Thanks for having a look at this pull request and your comments!

I have been testing with the latest version of `jupyter_client==8.0.3` which produces the issues I've listed above. I believe version 8.0.0 is where the swap over from `notebook` 6 to `nbclassic` occurs. I have been trying to regularly rebuild my python environments and keep up with the latest versions of packages wherever possible. From my testing the current version of appmode works just fine if you pin:
```
jupyter_client<8.0.0
notebook<=6.4.11
```
Not 100% sure what you mean by backend, I have done some testing with multiple versions of `jupyter_client`, using either `notebook` 6 or `nbclassic`. The current working environment I am running uses the above version pins to ensure appmode works, which is notebook 6 based. This uses the `jupyter notebook` command line entry point. In testing I've also been trying the `jupyter lab` entry point and then swapping over to `/tree` to access notebook 6 or `nbclassic` notebooks. Ideally I want to run a jupyter lab environment and also have the ability to serve out appmode apps via the appropriate URLs. 

If the plan was to wait until notebook-v7 lands then I am happy to wait until then. As you say it probably makes more sense to try and support notebook-v7 development, rather than forking the notebook html templates, I'm just not sure how much work that would involve.

This fork serves the purpose I need for now, including this version in builds will resolve the current compatibility issues as a temporary fix until notebook v7 is released. Did you want me to close the pull request? Or are there any parts you think it would make sense to merge in now, like the `server_extension.py` changes? I could move the notebook templates to a branch to allow for a merge without them.

unkcpz commented:
Thanks for the detail, can you try to run following code in a notebook you start?

```
import psutil
[i for i in psutil.Process().parent().cmdline()]
```

In our AiiDAlab deployment, we need to set `JUPYTERHUB_SINGLEUSER_APP: "notebook.notebookapp.NotebookApp"` in the k8s config file to force using the NotebookApp backend. 
I think in the terminal, `jupyter notebook` command will start with the notebook mode, maybe it uses the `nbclassic` by default. I'll check in my environment.

> This fork serves the purpose I need for now, including this version in builds will resolve the current compatibility issues as a temporary fix until notebook v7 is released. Did you want me to close the pull request? Or are there any parts you think it would make sense to merge in now, like the server_extension.py changes? I could move the notebook templates to a branch to allow for a merge without them.

If you can pin the jupyter-client version, I advise to do it for now so you can use the App mode without isuses. I'll come back with the test if it possible to using the jupyter-client v8 with using the notebook backend and running appmode without issue. 
As for the asyncio/tornado issue, I think your change is anyway valuable. Could you separate it and make a new PR?

unkcpz commented:
I did a test which try to use minimal change to run the appmode in jupyter-client. You can find my change at https://github.com/oschuett/appmode/pull/66, simply as you mentioned, using the flag to allow create the hidden file. 
If you build the container and run the example, it is all fine. Did I miss something?

pdstack commented:
Thanks for your comment about the backend, while the psutil command didn't really tell me anything, the other link was quite informative. The terminal window running the jupyter command makes it easy to tell which backend is being used, the logging messages are prefixed with NotebookApp or ServerApp. 

As a result I realised what I am actually trying to acheive here is getting appmode to work with the ServerApp backend. While `jupyter_server<2.0.0` can run the `/lab` mode using the NotebookApp backend, newer `jupyter_server>=2.0.0` versions require the ServerApp backend to do so.

The issues with `jupyter_client>8.0.0` weren't really related to appmode specifically, but rather the NotebookApp backend, and interactivity with some of the apps I have been working on when using `jupyter_client>8.0.0`. Using ServerApp resolves these problems. See https://discourse.jupyter.org/t/jupyter-notebook-zmq-message-arrived-on-closed-channel-error/17869 for the error I was encountering. There is probably something else going on here, but I would prefer to use the ServerApp backend anyway.

I had copied the notebook html templates because they are not available in the ServerApp templates, I managed to get them working by using `--ServerApp.extra_template_paths='***site-packages***\notebook\templates'`, but this is obviously not an ideal solution. What I have done in the end is add them to the templates path in the `load_jupyter_server_extension` function, similar to how the current appmode templates are added. This is a more elegant solution than what I had done previously.

I have also added another configuration option: hidden_temp_files, which adds/removes the `.` at the start of the file name. I don't want to enable `ContentsManager.allow_hidden=True` in my environment, there a number of other hidden folders that are related to the container setup and python virtual environment which would clutter up the jupyter lab interface if shown. I have updated the readme to let people know they will have to either enable `ContentsManager.allow_hidden=True` or `Appmode.hidden_temp_files=False` for appmode to work with newer notebook versions. It would make more sense for me that `Appmode.hidden_temp_files` defaults to True, so no additional configuration is required for appmode to work, but I'm not sure if you have specific reasons to prefer using `ContentsManager.allow_hidden=True` ala #66, so I have left the default as False to maintain current behaviour.

I think this pull request is in a good spot now, and enabling ServerApp backend compatibility should make Appmode more usable generally going forward.

unkcpz commented on appmode/server_extension.py:
```diff
@@ -87,52 +99,62 @@ def get(self, path):
-
-        return(tmp_path)
+        pc = cm.copy(path, tmp_path)
+        if isinstance(pc, collections.abc.Awaitable):
```
Nice check here! I like it.

unkcpz replied:
Not sure it is easy to implement this pattern for all such changes. You use a lot of

```python
        pd = cm.delete(path)
        if isinstance(pd, collections.abc.Awaitable):
            await pd
```

Is that possible to add this to the `self.context_manager`?

pdstack replied:
The content_manager/session_manager is part of the underlying jupyter stack I believe. These checks are repetitive, but they are similar to what was there before, i.e 
```
        if gen.is_coroutine_function(sm.get_session):
            s = yield sm.get_session(path=path)
        else:
            s = sm.get_session(path=path)
        if gen.is_coroutine_function(sm.delete_session):
            yield sm.delete_session(session_id=s['id'])
        else:
            sm.delete_session(session_id=s['id'])
```
The session manager and content manager seem to have moved across to using the newer native async functionality instead of the older tornado coroutines, hence the change to checks on async. There may be a better or more canonical way to do these checks, I haven't seen any other examples, this is just what I came up with to do the same thing as the previous code was trying to do.

We could perhaps make a wrapper for it, like
```
def check_async(x):
    if isinstance(x, collections.abc.Awaitable):
        x = await x
    return x
````
Calls then become ```check_async(cm.delete(path))``` or ```s = check_async(sm.get_session(path=path))```. I can swap it over to doing that if you think it would be better?

Since these checks are for backwards compatibility, another option would be to remove them and just use the awaits, i.e.
```
s = await sm.get_session(path=path)
```
Would then need to version pin the dependency that changes over to async content and session managers.

pdstack replied:
Following the pattern in jupyter_core for handling functions that may or may not be async (see https://github.com/jupyter/jupyter_core/blob/main/jupyter_core/utils/__init__.py#L172), I've added a simple ensure_async function to wrap the contents manager and session manager calls.

This cleans up the repetitive checks for Awaitable objects, and I believe resolves this issue.

unkcpz reviewed (changes requested):
@pdstack thanks a lot for further investigation. I think the changes look all good to me. I just have a minor request on the duplicate pattern for all `await` clause changes. Let me know if it is possible to simplify in context manager. If it is to complex I am okay with the current change. I'll make a test after.

unkcpz commented on appmode/server_extension.py:
```diff
@@ -1,17 +1,29 @@
 from traitlets import Bool, Unicode
 
 
+async def ensure_async(obj):
```
The name is a bit confusing with the `asyncio.ensure_future`. I would suggest call the function `def await_if_awaitable(obj)`.

pdstack replied:
I've renamed it to await_if_awaitable. The name ensure_async was what jupyter_core used for the function, but I don't have any strong opinions on it.

unkcpz replied:
Okay, then I think they are all fine, since you already changed them, I won't ask you to change them back :smile_cat: . Thanks!

pdstack replied:
No worries, easy enough to change back if you think the other name was better, or you think of another one. Thanks for the quick reviews!

unkcpz reviewed (changes requested):
Thanks a lot for having a generic function for the await pattern @pdstack 
Just one minor request and then I think it is good to go.

unkcpz commented:
I test locally and it works neatly. :+1:

unkcpz commented:
Okay, merged. Thanks again @pdstack 

I plan to clean up the release flow and make a new release. Although it is backward compatible, maybe still good to make a minor version release to `0.9.0`? What do you think @pdstack @oschuett @yakutovicha?

pdstack commented:
I think a minor release would make sense, there are 2 configuration flags added, as well as python 2 support being dropped.
